File: 92-reverse-linked-list-ii/92-reverse-linked-list-ii.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int left, int right) {

        Deque<ListNode> stack = new ArrayDeque<>();

        ListNode leftNode = head;
        ListNode rightNode = head;

        for(int i = 0; i < right; i++) {
            if(i == left-1) leftNode = rightNode;
            stack.push(rightNode);
            rightNode = rightNode.next;
        }

        rightNode = stack.pop();
        while(true) {
            if(leftNode == rightNode || rightNode.next == leftNode) break;

            int t = leftNode.val;
            leftNode.val = rightNode.val;
            rightNode.val = t;

            rightNode = stack.pop();
            leftNode = leftNode.next;
        }

        return head;
    
    }
}

File: 92-reverse-linked-list-ii/92-reverse-linked-list-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        isStop = False
        leftNode = head
        def recurse(rightNode, m, n):
            nonlocal isStop, leftNode
            if n == 1:
                return
            rightNode = rightNode.next
            if m > 1:
                leftNode = leftNode.next
            recurse(rightNode, m-1, n-1)
            if leftNode == rightNode or rightNode.next == leftNode:
                isStop = True
            if not isStop:
                leftNode.val, rightNode.val = rightNode.val, leftNode.val
                leftNode = leftNode.next
        recurse(head, left, right)

        return head